### ISSUE TITLE
Different behaviour of Dart VM and dart2js

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#30813]
The following co19 tests fail on dart2js and dartk in a checked mode with a RuntimeError, but works well on VM
Language/Expressions/Function_Expressions/static_type_dynamic_async_t03.dart
Language/Expressions/Function_Expressions/static_type_dynamic_asyncs_t03.dart
Language/Expressions/Function_Expressions/static_type_dynamic_syncs_t03.dart
Language/Expressions/Function_Expressions/static_type_form_3_async_t03.dart
Language/Expressions/Function_Expressions/static_type_form_3_asyncs_t03.dart
Language/Expressions/Function_Expressions/static_type_form_3_syncs_t03.dart

Example
```dart
import 'dart:async';

typedef Future streamFuncParam(int p1, [bool p2]);

main() {
  streamFuncParam sfp1 = (int p1, [bool p2]) async* {}; /// static type warning in analyzer, no issue on VM (both checked and production mode), runtime error in a cheked mode on dartk and dart2js
  sfp1(1);
}
```
Please clarify, which behavior is correct

**Comment by lrhn** [comment]
We are hitting a corner of the specification that isn't fully specified.

The typedef says that the function returns a `Future`, even if its name and the function assigned to it actually returns a `Stream` (the function expression is marked `async*`).

The static type of the function literal is `Stream<dynamic> Function(int, [bool])`, and it's correct to give a static warning, that function type is not assignableto one returning `Future<dynamic>`.
The correct static behavior is irrelevant to the VM, since it doesn't do static semantics.

The *runtime type* of the function literal is not specified anywhere. Its static type returns a `Stream<dynamic>`, but the specification is completely mute on the runtime semantics of function literals.

The VM implements `(int p1, [bool p2] async* {}` as a function object with runtime type `dynamic Function(int, [bool])`. Nothing says it can't, and with that runtime type, it's behavior is correct.

Dart2js implemenst it as a function object with runtime type `Stream<dynamic> Function(int, [bool])`, and that too is correct, since nothing is specified, nothing can be wrong.

In the longer run (Dart 2), the function literal *will* get a return type, either inferred from the surrounding or deduced to be at least `Stream<dynamic>`, but for now, the VM is technically correct.

**Comment by aam** [comment]
> The static type of the function literal is Stream<dynamic> Function(int, [bool]), and it's correct to give a static warning, that function type is not assignableto one returning Future<dynamic>.
> The correct static behavior is irrelevant to the VM, since it doesn't do static semantics.

Unless VM runs in checked mode, right, in which case it has to throw on type mismatch?

**Comment by lrhn** [comment]
Checked mode is not *static* checking, so the static type of the expression isn't relevant, only the runtime class of value the expression evaluates to. 
In this case, the VM's function's runtime type is a subtype of the variable type because the former uses `dynamic` as return type, and in Dart 1 that can never be wrong.

**Comment by aam** [comment]
Ah, interesting. Is it reasonable expectation for the spec to spell out runtime type of a literal, with assumption that it is different from static type? Are there other parts of the specs where we do that?

**Comment by lrhn** [comment]
The runtime type of a *literal* is generally not different from its static type. 

Functions were always special because `(x)=>e` has a *static* return type that is the static type of `e` (it's really the one instance of "type inference" in Dart 1), but since the VM doesn't contain a static type system at all, it cannot possibly make that the runtime type. It just doesn't know what the static type of `e` is. So, from the beginning, the function literals were special in this way. The spec didn't say what the runtime type was. It was an accidental omission, but since everybody agreed, nobody really noticed.
Function expressions with `{block}` bodies had `dynamic` as static return type, so the VM just used `dynamic` as return type for all function expressions, and `dynamic` always works, so nobody noticed.
I'd even say that the VM was correct.

When we added async functions, more static types were added, but still no runtime semantics. This made the current problem possible - the VM kept using `dynamic` as return type of all function expressions, other implementations chose to match the static type, at least one level deep, to return `Future<dynamic>`, `Stream<dynamic>` or `Iterable<dynamic>` from an `async`, `async*` or `sync*` function respectively.
That's what the current issue is about, a visibly different behavior because of different runtime types. When used correctly, you probably wouldn't notice, but here the function is being used incorrectly, and the VM's `dynamic` runtime-return type cannot detect that.

As we move towards Dart 2, the differences should disappear. We do need to specify the runtime type of the value of function expressions before we get there, likely just to match the current strong-mode behavior.
